File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2022/Day1.kt
package io.github.ocirne.aoc.year2022

import io.github.ocirne.aoc.AocChallenge

class Day1(val lines: List<String>) : AocChallenge(2022, 1) {

    override fun part1(): Int {
        return sumCalories().maxOrNull()!!
    }

    override fun part2(): Int {
        return sumCalories().sorted().reversed().take(3).sum()
    }

    private fun sumCalories(): List<Int> {
        val indexOfNewLines = listOf(-1) +
                lines.mapIndexedNotNull { index, s -> index.takeIf { s.isEmpty() } } +
                listOf(lines.size)
        return indexOfNewLines.zipWithNext { fromIndex, toIndex ->
            val calories = lines.subList(fromIndex + 1, toIndex)
            calories.sumOf { i -> i.trim().toInt() }
        }
    }
}

File: python/src/aoc/year2022/day1.py
from aoc.util import load_input, load_example


def calorieSums(lines):
    s = 0
    for line in lines:
        try:
            s += int(line)
        except ValueError:
            yield s
            s = 0
    yield s


def part1(lines):
    """
    >>> part1(load_example(__file__, "1"))
    24000
    """
    return max(calorieSums(lines))


def part2(lines):
    """
    >>> part2(load_example(__file__, "1"))
    45000
    """
    return sum(sorted(calorieSums(lines))[-3:])


if __name__ == "__main__":
    data = load_input(__file__, 2022, "1")
    print(part1(data))
    print(part2(data))
